### DISCUSSION TITLE
desktop app - retrieve absolute file path from VFileInput

### DISCUSSION THREAD

**Discussion by Zibc** [Kitware/trame#480] [Q&A] [answered]
Hello,
I discovered trame a while ago and finally decided to try something with it: a simple desktop app, not even using vtk.
Since i am only trying to evaluate how the whole thing work, I have now a very stupid example that has an interactive slider, and a file input field.
Using pywebview to render the ui.

I now face a potentially classic web-app issue: my goal with the file input is only to select a file to get the local file path, not to upload anything to the server side.
I now get the file details when selecting the file, but I am wondering whether it is possible to access the local file path instead of the file name and its content only.

I saw that it seems to be possible for pywebview to get access to the local file system, it seems like the same is also possible using Tauri, but I have no idea whether such possibilities are exposed in trame, or if they could achieved. (electron apps also allow local file system access to my knowledge, but I never did anything with electron, so i essentially repeating what I read online here)

A hint would be appreciated

To recall my goal: build a simple desktop app in which I can input local file path for later processing on the python side: the app does not need to process the content of file (it will utimately copy it else where, but that's another problem).

Thanks in advance.

**Comment by jourdain** [comment]
The better path to deal with file dialog and get the path, is with Tauri. You can find [this example that showcases the file dialog and more](https://github.com/Kitware/trame-tauri/tree/master/examples/03_tauri_ws_menu_dialogs).

**Comment by Zibc** [comment]
Thanks for pointing in this direction.
I managed to compile the example. (although surprised by the quantity of dev libraries I had to install to get it to compile)
If I follow your reasoning, your suggestion is to create a button that triggers tauri to show a dialog and get the callback on the python side ?
If that's the case, getting the click to trigger the python callback: easy, but in this python callback, how do I ask tauri to show the dialog ?
can I instantiate tauri?Dialog().open(...) directly from the python callback, or is it intended to be used only in the _build_ui context ?

**Reply by jourdain** [reply]
While you can do that, you can also use the utils in your template with something like below (which I did not try but should work the same.

```python
state.file_to_open = None

@state.change("file_to_open")
def open_file(file_to_open, **kwargs):
     if file_to_open is None:
         return
     # ... do your thing with the path

html.Button("Open", click="file_to_open = await utils.tauri.dialog.open()")
```

**Reply by Zibc** [reply]
Thanks for the tip! That makes it more consistent with how other field are processed!
I haven't tried yet, but I will soon.
Thanks for the help.

**Reply by Zibc** [reply]
To come back on this, I finally took the take to try a bit more. I cannot get this solution to work though. I only get a "white screen" in tauri if I use "click="file_to_open = await utils.tauri.dialog.open()" in the button definition.

So far, the solution I found is still based on my previous idea:
```python
from trame.app import get_server, asynchronous
from trame.ui.vuetify3 import SinglePageLayout
from trame.widgets import vuetify3 as vuetify, tauri
from trame.decorators import TrameApp, life_cycle


@TrameApp()
class Cone:
    def __init__(self, server=None):
        self.server = get_server(server)
        self.ui = self._build_ui()
        self.state.file_to_open = None

    @property
    def state(self):
        return self.server.state

    @property
    def ctrl(self):
        return self.server.controller

    # -----------------------------------------------------
    # Dialog interactions
    # -----------------------------------------------------

    async def dialog_open(self):
        with self.state:
            self.state.logs += 'Open\n'

        with self.state:
            response = await self.ctrl.open('Open')
            self.state.logs += f'=> {response}\n'
            self.state.file_to_open = response

    @life_cycle.server_ready
    def _tauri_ready(self, **_):
        print(f'tauri-server-port={self.server.port}', flush=True)

    @life_cycle.client_connected
    def _tauri_show(self, **_):
        print('tauri-client-ready', flush=True)

    def choose(self):
        asynchronous.create_task(self.dialog_open())

    def _build_ui(self):
        with SinglePageLayout(self.server) as layout:
            self.ctrl.open = tauri.Dialog().open
            # with tauri.Dialog() as dialog:
            #     self.ctrl.open = dialog.open
            # Toolbar
            with layout.toolbar.clear() as toolbar:
                toolbar.density = "compact"
                vuetify.VToolbarTitle("my test trame app")

            # Main content
            with layout.content:
                with vuetify.VContainer(fluid=True):
                    with vuetify.VRow(align="stretch"):
                        with vuetify.VCol(cols=9):
                            vuetify.VTextField(label='None given',
                                               v_model=('file_to_open', None),)
                        with vuetify.VCol(cols=3):
                            vuetify.VBtn('Browse', click=self.choose)

                    with vuetify.VRow(align="stretch"):
                        with vuetify.VCol(cols=12):
                            vuetify.VTextarea(v_model=('logs', ''), readonly=True)

            return layout


if __name__ == '__main__':
    app = Cone()
    app.server.start()
```

And I am wondering whether the code I wrote is even actually good (it does work, but I am not too sure about the why and how)
If you have next hint, that would be very appreciated!

**Comment by jourdain** [comment, marked as answer]
You may have to do `utils.tauri.dialog.open().then((v) => {file_to_open = v;}))`. `await` may not be allowed in inlined function within a template.

**Reply by Zibc** [reply]
That worked out perfectly!
Thanks for your quick reaction!
